tensorflow_addons/layers/tests/crf_test.py

```diff
@@ -0,0 +1,329 @@
+        chain_initializer=tf.keras.initializers.Constant(transitions),
+        use_boundary=True,
+        boundary_initializer=tf.keras.initializers.Constant(boundary_values),
+        name="crf_layer",
+    )
+    return logits, tags, transitions, boundary_values, crf_layer
+
+
+def test_keras_model_inference():
+    logits, _, _, _, crf_layer = get_test_data_extended()
+
+    input_tensor = tf.keras.layers.Input(shape=(3, 5))
+    decoded_sequence, _, _, _ = crf_layer(input_tensor)
+    model = tf.keras.Model(input_tensor, decoded_sequence)
+
+    model.predict(logits)
+    model(logits).numpy()
+
+
+def test_unmasked_viterbi_decode():
+
+    x_np, y_np = get_test_data()
+
+    transitions = np.ones([5, 5])
+    boundary_value = np.ones(5)
+
+    layer = CRF(
+        units=5,
+        use_kernel=False,  # disable kernel transform
+        chain_initializer=tf.keras.initializers.Constant(transitions),
+        use_boundary=True,
+        boundary_initializer=tf.keras.initializers.Constant(boundary_value),
+    )
+
+    decoded_sequence, _, _, _ = layer(x_np)
+    decoded_sequence = decoded_sequence.numpy()
+    np.testing.assert_equal(decoded_sequence, y_np)
+    assert decoded_sequence.dtype == np.int32
+
+
+def unpack_data(data):
+    if len(data) == 2:
+        return data[0], data[1], None
+    elif len(data) == 3:
+        return data
+    else:
+        raise TypeError("Expected data to be a tuple of size 2 or 3.")
+
+
+class ModelWithCRFLoss(tf.keras.Model):
+    """Wrapper around the base model for custom training logic."""
+
+    def __init__(self, base_model):
+        super().__init__()
+        self.base_model = base_model
+
+    def call(self, inputs):
+        return self.base_model(inputs)
+
+    def compute_loss(self, x, y, sample_weights, training=False):
+        y_pred = self(x, training=training)
+        _, potentials, sequence_length, chain_kernel = y_pred
+
+        # we now add the CRF loss:
+        crf_loss = -crf_log_likelihood(potentials, y, sequence_length, chain_kernel)[0]
+
+        if sample_weights is not None:
+            crf_loss = crf_loss * sample_weights
+
+        return tf.reduce_mean(crf_loss), sum(self.losses)
+
+    def train_step(self, data):
+        x, y, sample_weight = unpack_data(data)
+
+        with tf.GradientTape() as tape:
+            crf_loss, internal_losses = self.compute_loss(
+                x, y, sample_weight, training=True
+            )
+            total_loss = crf_loss + internal_losses
+
+        gradients = tape.gradient(total_loss, self.trainable_variables)
+        self.optimizer.apply_gradients(zip(gradients, self.trainable_variables))
+
+        return {"crf_loss": crf_loss, "internal_losses": internal_losses}
+
+    def test_step(self, data):
+        x, y, sample_weight = unpack_data(data)
+        crf_loss, internal_losses = self.compute_loss(x, y, sample_weight)
+        return {"crf_loss_val": crf_loss, "internal_losses_val": internal_losses}
```

Comment:
prefix `val_` is [already added](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/training.py#L930)